distributed/deploy/local_env.py

```diff
@@ -0,0 +1,279 @@
+        self.kwargs = kwargs
+        self.worker_module = worker_module
+        self.name = name
+
+    async def start(self):
+        set_env = await _set_env_helper(self.connect_options)
+
+        cmd = " ".join(
+            [
+                set_env,
+                self.python_exe,
+                "-m",
+                self.worker_module,
+                self.scheduler,
+                "--name",
+                str(self.name),
+            ]
+            + cli_keywords(self.kwargs, cls=_Worker, cmd=self.worker_module)
+        )
+        self.proc = await asyncio.create_subprocess_shell(
+            cmd, stderr=asyncio.subprocess.PIPE, **self.connect_options
+        )
+
+        # We watch stderr in order to get the address, then we return
+        while True:
+            line = await self.proc.stderr.readline()
+            if not line.decode("ascii").strip():
+                raise Exception("Worker failed to start")
+            else:
+                line = line.decode("ascii").strip()
+            logger.info(line)
+            if "worker at" in line:
+                self.address = line.split("worker at:")[1].strip()
+                self.status = Status.running
+                break
+        logger.debug("%s", line)
+        await super().start()
+
+
+class Scheduler(Process):
+    """A Remote Dask Scheduler run by a specified Python executable
+
+    Parameters
+    ----------
+    python_exe: str
+        Full path to Python executable to run this scheduler
+    connect_options: dict
+        kwargs to be passed to asyncio subprocess connections
+    kwargs: dict
+        These will be passed through the dask-scheduler CLI to the
+        dask.distributed.Scheduler class
+    """
+
+    def __init__(self, python_exe: str, connect_options: dict, kwargs: dict):
+        super().__init__()
+
+        self.python_exe = python_exe
+        self.kwargs = kwargs
+        self.connect_options = connect_options
+
+    async def start(self):
+        logger.debug("Created Scheduler")
+
+        set_env = await _set_env_helper(self.connect_options)
+
+        cmd = " ".join(
+            [set_env, self.python_exe, "-m", "distributed.cli.dask_scheduler"]
+            + cli_keywords(self.kwargs, cls=_Scheduler)
+        )
+        self.proc = await asyncio.create_subprocess_shell(
+            cmd, stderr=asyncio.subprocess.PIPE, **self.connect_options
+        )
+
+        # We watch stderr in order to get the address, then we return
+        while True:
+            line = await self.proc.stderr.readline()
+            if not line.decode("ascii").strip():
+                raise Exception("Scheduler failed to start")
+            else:
+                line = line.decode("ascii").strip()
+            logger.info(line)
+            if "Scheduler at" in line:
+                self.address = line.split("Scheduler at:")[1].strip()
+                break
+        logger.debug("%s", line)
```

Comment:
This block here looks almost identical to the section in `Scheduler`. Would it make sense to bump some of this up into the super class?

Reply:
I considered this - doing something like I did with the environment setter helper. I ended up not doing so because it's not completely identical (the strings 'Scheduler' and 'Worker' change, and the specific search string for locating the addresses) which I thought might lead to a proliferation of kwargs. Also, appealing to precedent, the `SSHCluster` implementation has quite a bit of duplicated code 😉 

I do like avoiding duplicated code though, so happy to try it out. Want me to also integrate that separate helper function (`_set_env_helper`) into the superclass?

Reply:
Sure